magento/pwa-studio
# Service Worker HTML caching bug fix

## Description

This PR addresses the issue of Venia based apps showing white screen (crashing) when a new version is released.

### Background
The Service Worker provided as part apps built using the `PWA-Studio` are built using the workbox framework. Workbox provides a bunch of utilities to ease our work of SW development/maintenance. Service worker files are built along with the client code using webpack during the build stage but are not loaded when the app loads.

Here is a timeline of what will happen when a user loads the app the first time.

```
Venia Based App Load
           |__ index.html requested
                      |__ client.************.js requested
                                 |__ Service Worker Registration
                                            |__ Service Worker Activated
                      |__ runtime.++++++++.js requested
                      |__ vendors.-----------.js requested
```

As shown above, the Service Worker is loaded and activated after the JS files are requested and loaded. Hence the first time when this happens the Service Worker was not able to handle those requests and cache their responses.

But, if the user refreshes the page or goes to a new URL in the same domain since the SW is active, it will handle those requests and if the requests have been cached in the storage, it will return them immediately instead of fetching them from the server.

The SW also caches the `index.html` file as well. `index.html` file is requested every time the user refreshes or navigates to a new URL but it is not changed often. When the file is requested by the app, the SW will return the caches HTML file from the cache and in the background request the new HTML file. After the request has been completed, the SW checks the files for changes and if it determines that the file has changed (one possibility being a new version of the app has been deployed), it sends a message to the app to show a toast asking the user to refresh.

### Problem
As mentioned above, the SW caches the HTML file for quick access. If it happens that, when the file has been returned, if it has links that are not available anymore either in cache or on the remote server, the app crashes because the network would resolve with a 404 error.

This happens to be a super edge case. This only happens when the SW has not cached all the JS files but has cached the HTML file and a new deployment is available, hence even the server does not have files anymore. This wouldn't happen if the user refreshes the page twice / gives enough time for the SW to cache the JS files.

### Steps to Reproduce
@Jordaneisenburger found the bug and has documented the steps in this [video](https://www.youtube.com/watch?v=IzNF-vv4Kow).

To test it out locally:
1. Checkout `develop` version of `pwa-studio`.
2. Build and load the app as shown in the video in different terminals. Terminal 1: `yarn build` Terminal 2: `yarn stage:venia`.
3. Load the app in a new browser instance. Make sure the cache is clean.
4. After the app has loaded, refresh again. This will make sure the SW has cached all the files. In the first load, the SW is loaded and activated, and in the second load/refresh it has cached all the files.
5. Check the cache storage section in the dev tools, you should see a cache storage instance called `workbox-runtime`. It should have the HTML and JS files cached by the SW.
6. Click on the Client JS file, and delete it from the cache.
7. Now make a change in the code, as shown in the video, and perform `step 2`. This should create a new `client.[HASH].js` file.
8. Refresh the app. It should crash because the HTML file returned by the SW has reference to the old `client.[HASH].js` which is not in the cache or in the remote server.

These steps replicate a new version release because every time we release a new version, we are releasing a new version of our app's codebase which stays in `client.[HASH].js` (parts of it will be in vendors and runtime as well).

### Solution
Every time the app requests for the HTML file, instead of returning the caches HTML file, the SW will perform a validation test, which checks for the script files in the cache. If the script files are not cached, the SW will not return the cached HTML file, instead, it will fetch the new HTML, cache it and then return it to the browser. This will ensure that the SW will never return a file that has invalid references.

This will not cause any performance hit, because the validation will fail only the first time the app requests the HTML file in the SW's lifetime. From next time onwards, the SW would have cached the relevant files and the validation will pass.

To validate this, perform the same set of steps as mentioned above on this branch.

## Related Issue
Closes [PWA-408](https://jira.corp.magento.com/browse/PWA-408)
Closes #2351 

## Acceptance 
The app should not crash on a new version release.

### Verification Stakeholders
@zetlen 
@jimbo 
@Jordaneisenburger 

### Verification Steps
1. Make sure the cache is cleared and load the [deployed app](https://pr-2376.pwa-venia.com/).
2. Refresh it again to make sure all the JS files have been cached.
3. Delete the client js file from the `workbox-runtime` cache.
4. Load the app again, instead of crashing the SW should fetch the missing client js file and load the app.

You can also perform a sophisticated check like mentioned in the description, but unfortunately, it can only be done locally because it involves code changes and rebuilds.

## Checklist
* Should not change anything in `dev` mode. This is only intended for the `stage` mode because the SW is not active in the `dev/watch` mode.
